Add support for contract upgrades

patrickkuo opened the pull request:
* Add interface for the upgraded contract to implement, which provides functionality for upgrading legacy states.
* Add shared upgrade command and verification code for it.
* Add DummyContractV2 to illustrate what an upgraded contract looks like.
* Add new functions to vault service to support upgrading state objects.
* Add contract upgrade flow

mikehearn commented on core/src/main/kotlin/net/corda/flows/AbstractStateReplacementFlow.kt:
```diff
@@ -24,19 +24,23 @@ import net.corda.flows.AbstractStateReplacementFlow.Instigator
+     * @param S the input contract state
+     * @param T the output contract state, this could be different from [S] in some cases. For example, in contract upgrade, the output state object might be different from the input state.
+     */
+    abstract class Instigator<out S : ContractState, out T : ContractState, M>(val originalState: StateAndRef<S>,
```
You've added M but it's not documented. Also, can move type vars out like it was before.

mikehearn commented on core/src/main/kotlin/net/corda/flows/ContractUpgradeFlow.kt:
```diff
@@ -0,0 +1,75 @@
+            val state = tx.tx.outRef<ContractState>(proposal.stateRef.index)
+            val acceptedUpgrade = serviceHub.vaultService.getAcceptedContractStateUpgrade(state) ?: throw IllegalStateException("Contract state upgrade rejected. State hash : ${state.ref}")
+            val actualTx = proposal.stx.tx
+            actualTx.inputs.map { serviceHub.vaultService.statesForRef(it) }
```
Is this just a way to check that the transaction is resolved? Why not invoke ResolveTransactionsFlow instead?

mikehearn commented on core/src/main/kotlin/net/corda/flows/ContractUpgradeFlow.kt:
```diff
@@ -0,0 +1,75 @@
+class ContractUpgradeFlow<S : ContractState, out T : ContractState, out U : ContractUpgrade<S, T>>(val state: StateAndRef<S>,
+                                                                                                   val newContract: U) : FlowLogic<ContractUpgradeResponse>() {
+    @Suspendable
+    override fun call() = try {
```
body func

mikehearn commented on core/src/main/kotlin/net/corda/flows/ContractUpgradeFlow.kt:
```diff
@@ -0,0 +1,75 @@
+    }
+}
+
+class ContractUpgradeFlow<S : ContractState, out T : ContractState, out U : ContractUpgrade<S, T>>(val state: StateAndRef<S>,
```
These flows need documenting and maybe grouping. It's rather confusing to read all this.

mikehearn commented on core/src/main/kotlin/net/corda/flows/ContractUpgradeFlow.kt:
```diff
@@ -0,0 +1,75 @@
+    override fun call() = try {
+        ContractUpgradeResponse.Accepted(subFlow(ContractUpgradeReplacementFlow.InstigatorFlow(state, newContract)))
+    } catch (e: Exception) {
+        ContractUpgradeResponse.Rejected(e.message)
```
Please just let things grind to a halt now in case of error. The more cases like this we introduce, the more cases Shams has to delete when exception handling in the flow framework is done.

mikehearn commented on node/src/main/kotlin/net/corda/node/services/ContractUpgradeService.kt:
```diff
@@ -0,0 +1,19 @@
+import net.corda.flows.ContractUpgradeReplacementFlow
+import java.util.function.Function
+
+object ContractUpgrade {
```
This stuff is core functionality, it shouldn't be a plugin. Please register the flows in the node setup code instead.

rnicoll commented on core/src/main/kotlin/net/corda/core/contracts/clauses/UpgradeClause.kt:
```diff
@@ -0,0 +1,40 @@
+
+            val upgradeContract = command.value.contractUpgrade
+
+            for (stateIdx in 0..inputs.size - 1) {
```
Would https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/for-each-indexed.html perhaps be neater here?

patrickkuo commented:
Hi @mikehearn  @rnicoll, thanks for the review.

I have added more doc / comments to the code and also a page for the docsite.
I have also tried to simplify the code and reduce use of type variables where possible.
Most of the coding style issues had been addressed.

I think this PR is ready for review again, thanks!

mikehearn commented on core/src/main/kotlin/net/corda/flows/ContractUpgradeFlow.kt:
```diff
@@ -0,0 +1,67 @@
+        @Suspendable
+        override fun verifyProposal(maybeProposal: UntrustworthyData<AbstractStateReplacementFlow.Proposal<UpgradedContract<ContractState, ContractState>>>): AbstractStateReplacementFlow.Proposal<UpgradedContract<ContractState, ContractState>> {
+            return maybeProposal.unwrap { proposal ->
+                val stx = serviceHub.storageService.validatedTransactions.getTransaction(proposal.stateRef.txhash) ?: throw IllegalStateException("We don't have a copy of the referenced state")
```
Let's add a brief comment here discussing why we directly fetch and fail if missing, rather than resolve the transaction to ensure it's there.

mikehearn commented on core/src/main/kotlin/net/corda/flows/ContractUpgradeFlow.kt:
```diff
@@ -0,0 +1,67 @@
+                val actualTx = proposal.stx.tx
+                val expectedTx = assembleBareTx(state, proposal.modification).toWireTransaction()
+                requireThat {
+                    "the instigator is one of the participants" by (state.state.data.participants.contains(otherSide.owningKey))
```
you can write this without the brackets as `a in b`

mikehearn commented on core/src/main/kotlin/net/corda/flows/ContractUpgradeFlow.kt:
```diff
@@ -0,0 +1,67 @@
+
+    class Acceptor(otherSide: Party) : AbstractStateReplacementFlow.Acceptor<UpgradedContract<ContractState, ContractState>>(otherSide) {
+        @Suspendable
+        override fun verifyProposal(maybeProposal: UntrustworthyData<AbstractStateReplacementFlow.Proposal<UpgradedContract<ContractState, ContractState>>>): AbstractStateReplacementFlow.Proposal<UpgradedContract<ContractState, ContractState>> {
```
Shouldn't the type here just be `Proposal`?

mikehearn commented on core/src/main/kotlin/net/corda/flows/ContractUpgradeFlow.kt:
```diff
@@ -0,0 +1,67 @@
+                val expectedTx = assembleBareTx(state, proposal.modification).toWireTransaction()
+                requireThat {
+                    "the instigator is one of the participants" by (state.state.data.participants.contains(otherSide.owningKey))
+                    "the proposed upgrade ${proposal.modification} is a trusted upgrade path" by (proposal.modification.javaClass == authorisedUpgrade.javaClass)
```
This will run the toString method of proposal.modification - I guess there's no risk that this will be an unexpected object that might do something weird?

mikehearn commented on core/src/main/kotlin/net/corda/flows/ContractUpgradeFlow.kt:
```diff
@@ -0,0 +1,67 @@
+                    "the proposed tx matches the expected tx for this upgrade" by (actualTx == expectedTx)
+                    "number of inputs and outputs match" by (proposal.stx.tx.inputs.size == proposal.stx.tx.outputs.size)
+                    "input belongs to the legacy contract" by (state.state.data.contract.javaClass == proposal.modification.legacyContract.javaClass)
+                    "output belongs to the upgraded contract" by proposal.stx.tx.outputs.all { it.data.contract.javaClass == proposal.modification.javaClass }
```
all? Why isn't this insisting that there is only one replacement state, if there is only one input state?

mikehearn commented on core/src/test/kotlin/net/corda/core/contracts/DummyContractV2Tests.kt:
```diff
@@ -0,0 +1,31 @@
+
+        assertEquals(v1Ref, tx.inputs.single())
+
+        val expectedOutput: TransactionState<ContractState> = TransactionState(contractUpgrade.upgrade(v1State.data), DUMMY_NOTARY)
```
I guess you don't need the type hint here, it doesn't add much as the initialiser on the right hand side is clear enough.

mikehearn commented on core/src/test/kotlin/net/corda/core/flows/ContractUpgradeFlowTest.kt:
```diff
@@ -0,0 +1,182 @@
+        // The request is expected to be rejected because party B haven't authorise the upgrade yet.
+        val rejectedFuture = a.services.startFlow(ContractUpgradeFlow.Instigator(atx!!.tx.outRef(0), DUMMY_V2_PROGRAM_ID)).resultFuture
+        mockNet.runNetwork()
+        assertFails { rejectedFuture.get() }
```
assertFailsWith?

mikehearn commented on core/src/test/kotlin/net/corda/core/flows/ContractUpgradeFlowTest.kt:
```diff
@@ -0,0 +1,182 @@
+        // Party B authorise the contract state upgrade.
+        b.services.vaultService.authoriseContractUpgrade(btx!!.tx.outRef<ContractState>(0), DUMMY_V2_PROGRAM_ID)
+
+        // Party A initiate contract upgrade flow, expected to success this time.
```
initiates, expected to succeed

mikehearn commented on core/src/test/kotlin/net/corda/core/flows/ContractUpgradeFlowTest.kt:
```diff
@@ -0,0 +1,182 @@
+        mockNet.runNetwork()
+        val stateAndRef = result.get().let {
+            when (it) {
+                is CashFlowResult.Success -> it.transaction?.tx?.outRef<Cash.State>(0)
```
I think this needs to be rebased now Shams' error work is in.

mikehearn commented on core/src/test/kotlin/net/corda/core/flows/ContractUpgradeFlowTest.kt:
```diff
@@ -0,0 +1,182 @@
+        // Starts contract upgrade flow.
+        a.services.startFlow(ContractUpgradeFlow.Instigator(stateAndRef!!, CashV2()))
+        mockNet.runNetwork()
+        // Get contract state form the vault.
```
typo: form -> from

mikehearn commented on core/src/test/kotlin/net/corda/core/flows/ContractUpgradeFlowTest.kt:
```diff
@@ -0,0 +1,182 @@
+        // Get contract state form the vault.
+        val state = databaseTransaction(a.database) { a.vault.currentVault.states }
+        assertTrue(state.size == 1)
+        assertTrue(state.first().state.data is CashV2.State, "Contract state is upgraded to the new version.")
```
could just be state.single() and the above assertion can then be skipped

mikehearn commented on core/src/test/kotlin/net/corda/core/flows/ContractUpgradeFlowTest.kt:
```diff
@@ -0,0 +1,182 @@
+            override fun withNewOwner(newOwner: CompositeKey) = Pair(Cash.Commands.Move(), copy(owners = listOf(newOwner)))
+
+            /** Object Relational Mapping support. */
+            override fun generateMappedObject(schema: MappedSchema): PersistentState {
```
schemas are optional, right? Do you need this to check what the tests are trying to verify?

mikehearn commented on docs/source/contract-upgrade.rst:
```diff
@@ -0,0 +1,58 @@
+Upgrading Contracts
+===================
+While every care is taken in development of contract code, 
```
Needs blank line after === I think

mikehearn commented on docs/source/contract-upgrade.rst:
```diff
@@ -0,0 +1,58 @@
+Upgrading Contracts
+===================
+While every care is taken in development of contract code, 
+inevitably upgrades will be required to fix bugs (in either of design or implementation). 
```
"in either design or implementation"

mikehearn commented on docs/source/contract-upgrade.rst:
```diff
@@ -0,0 +1,58 @@
+
+4. The developer of contract X discovers a bug in the contract code, and releases a new version, contract Y.
+
+5. At this point of time all nodes should stop issuing contract X.
```
"issuing states of contract X", this can be part of step 4 I think.

mikehearn commented on docs/source/contract-upgrade.rst:
```diff
@@ -0,0 +1,58 @@
+
+3. Time passes.
+
+4. The developer of contract X discovers a bug in the contract code, and releases a new version, contract Y.
```
They notify the users e.g. via a mailing list or some sort of app store.

mikehearn commented on docs/source/contract-upgrade.rst:
```diff
@@ -0,0 +1,58 @@
+
+Workflow
+--------
+The current design for how upgrades would work looks like this:
```
We can just say, "Here's the workflow for contract upgrades" - it's hopefully clear already to the reader that the information is current.

mikehearn commented on docs/source/contract-upgrade.rst:
```diff
@@ -0,0 +1,58 @@
+
+7. Banks A and B instruct their Corda nodes (via RPC) to be willing to upgrade state objects of contract X, to state objects for contract Y using agreed upgrade path.
+
+8. One of the parties [Instigator] initiates an reissue of state objects referring to contract X, to a new state object referring to contract Y.
```
reissue -> upgrade ?

mikehearn commented on docs/source/contract-upgrade.rst:
```diff
@@ -0,0 +1,58 @@
+
+8. One of the parties [Instigator] initiates an reissue of state objects referring to contract X, to a new state object referring to contract Y.
+
+9. A proposed transaction [Proposal], taking in the old state and outputting the reissued version, is created and signed with the node's private key.
```
This isn't KDoc so square brackets won't work here.

mikehearn commented on docs/source/contract-upgrade.rst:
```diff
@@ -0,0 +1,58 @@
+Authorising upgrade
+-------------------
+Each of the participants in the upgrading contract will have to instruct their node that they are willing to upgrade the state object before the upgrade.
+Currently the vault service is used to manage the authorisation records. User will be able to use RPC to perform such instructions.
```
"User will be able to use" -> "The administrator can use"

mikehearn commented on node/src/main/kotlin/net/corda/node/services/vault/NodeVaultService.kt:
```diff
@@ -344,14 +345,21 @@ class NodeVaultService(private val services: ServiceHub) : SingletonSerializeAsT
+
+    @Suppress("UNCHECKED_CAST")
+    override fun authoriseContractUpgrade(stateAndRef: StateAndRef<*>, upgrade: UpgradedContract<*, *>) {
+        "The upgraded contract is an upgrade of the contract state's contract." by (upgrade.legacyContract.javaClass == stateAndRef.state.data.contract.javaClass)
```
This won't do anything as it's not in a requireThat section! Actually I'm surprised it compiles. Why does it compile?

You can just write this in the old fashioned way: `if (!condition) throw IllegalStateException("...")`

Also this error message is kind of confusing!

patrickkuo commented on core/src/main/kotlin/net/corda/core/contracts/Structures.kt:
```diff
@@ -396,6 +396,24 @@ interface NetCommand : CommandData {
 }
 
+/** Indicates that this transaction replaces the inputs contract state to another contract state */
+class UpgradeCommand<in OldState : ContractState, out NewState : ContractState>(val upgradedContract: UpgradedContract<OldState, NewState>) : CommandData{
```
Agree, in the command we only checks equality of the UpgradeContract's class, passing in the instance is not necessary. I am also removing the type parameters as well as you rightly points out its mostly ignored.

patrickkuo commented on core/src/main/kotlin/net/corda/core/contracts/Structures.kt:
```diff
@@ -445,6 +463,29 @@ interface Contract {
+/**
+ * Convenience method for upgrading a contract state from state and reference.
+ */
+fun <S : ContractState, T : ContractState> UpgradedContract<S, T>.upgrade(ref: StateAndRef<S>): T = upgrade(ref.state.data)
```
This was left over code from previous check in and its not being used anywhere, it will be removed.

patrickkuo commented:
Using Class<UpdatedContract> in contract upgrade instead of instance of the contract. 
Extracted out verification logics to ContractUpgradeVerifier and not using Clauses. 
Also added more docs to the api, hopefully its now easier for the developers to understand how to start a contract upgrade.

@mikehearn I think its ready for more reviews, thanks!

mikehearn commented on core/src/main/kotlin/net/corda/flows/ContractUpgradeFlow.kt:
```diff
@@ -0,0 +1,83 @@
+    }
+}
+
+object ContractUpgradeVerifier {
```
Do these need to be in a singleton? Why not just put them in the Contract class (as static methods)?

mikehearn commented:
Needs a rebase. Other than a minor quibble about where to put the snippets of verification logic, I think we're done!

patrickkuo commented:
Thanks! @mikehearn
I have moved the verification logic to UpgradedContract class and also did a rebase.

mikehearn commented on core/src/main/kotlin/net/corda/core/contracts/Structures.kt:
```diff
@@ -463,6 +463,26 @@ interface UpgradedContract<in OldState : ContractState, out NewState : ContractS
     fun upgrade(state: OldState): NewState
+
+    companion object {
+        fun verify(tx: TransactionForContract) {
```
These both need @JvmStatic for Java users, otherwise it'll be a weird API from Java.

patrickkuo replied:
hmm.... looks like cannot mark method as @JvmStatic if they are in interface's companion object, will have to move them somewhere else...

mikehearn commented:
Add the annotations and then it LGTM! Hurrah!

mikehearn commented:
Pull the trigger! Do it! 🎉
